app/mender.go

```diff
@@ -554,13 +548,107 @@ func spinEventLoop(c *ControlMapPool, to State, ctx *StateContext, controller Co
+	switch to.Transition() {
+	case ToArtifactReboot_Enter:
+		status = "pause_before_rebooting"
+	case ToArtifactCommit_Enter:
+		status = "pause_before_committing"
+	case ToArtifactInstall:
+		status = "pause_before_installing"
+	}
+	if status != reported_status {
+		// Upload the status to the deployments/ID/status endpoint
+		log.Debugf("Reporting substatus: %s", status)
+		err := reporter(status)
+		if err == nil {
+			reported_status = status
+		} else {
+			log.Error(err)
+		}
+	}
+	// Schedule a timer for the next update map event to
+	// fetch from the server
+	updateMapFromServer := time.After(
+		ctx.NextUpdateMapTime.Sub(time.Now()))
+	log.Debugf("Next update refresh from the server at: %s", ctx.NextUpdateMapTime)
+
+	// wait until further notice
+	log.Debug("Pausing the event loop")
+	select {
+	case <-c.Updates:
+		log.Debug("Client control map update")
+	case <-updateMapFromServer:
+		log.Debug("Refreshing the update control map from the server")
+		_, err := controller.CheckUpdate()
+		if err == client.ErrNoDeploymentAvailable {
+			next, _ := to.HandleError(ctx, controller,
+				NewTransientError(errors.New("The deployment was aborted from the server")))
+			return next, false
+		}
+		if err != nil {
+			log.Errorf("Failed to update the update control map from the server, %s", err.Error())
+			err := refetch(controller)
+			if err == client.ErrNoDeploymentAvailable {
+				next, _ := to.HandleError(ctx, controller,
+					NewTransientError(errors.New("The deployment was aborted from the server")))
+				return next, false
+			}
+		}
+	}
+	return to, true
+}
+
+func refetch(controller Controller) error {
+	attempts := 0
+	for {
+		wait, err := client.GetExponentialBackoffTime(
+			attempts,
+			controller.GetUpdatePollInterval())
+		log.Infof("Next Update Check attempt in %d seconds", wait)
+		if err == client.MaxRetriesExceededError {
+			log.Errorf("Failed to refetch the update control map in %d attempts, giving up", attempts)
+			return err
+		} else if err != nil {
+			log.Errorf("Unexpected error: %s", err.Error())
+			return err
+		}
+		time.Sleep(wait)
+		_, err = controller.CheckUpdate()
+		attempts += 1
+		if err == client.ErrNoDeploymentAvailable {
+			return err
+		}
+		if err == nil {
+			return nil
+		}
+		log.Errorf("Attempt %d: Failed to update the control map from the server: %s", attempts, err.Error())
+	}
+}
+
 func transitionState(to State, ctx *StateContext, c Controller) (State, bool) {
 	from := c.GetCurrentState()
 
 	log.Infof("State transition: %s [%s] -> %s [%s]",
 		from.Id(), from.Transition().String(),
 		to.Id(), to.Transition().String())
 
+	// Check if we need to update the control map from the server
+	updateState, ok := to.(Updater)
+	if ok {
+		ctx.NextUpdateMapTime = c.GetControlMapPool().
+			NextControlMapHalfTime(updateState.Update().ID)
+		if time.Now().Sub(ctx.NextUpdateMapTime) < 0 {
+			// Maps are magically updated in CheckUpdate
+			_, err := c.CheckUpdate()
+			if err != nil {
```

Comment:
Should you check for `ErrNoDeploymentAvailable` here as well?

Reply:
What do you suggest we do in the case of an aborted deployment here?

Reply:
I guess the same as for other types of failures, `return from.HandleError(...)`, with some log message printed to say why the failure happened.

Reply:
Oh, and should check the return code from `refetch` as well.